receiver/receivertest/contract_checker.go

```diff
@@ -0,0 +1,424 @@
+	scenarios := []struct {
+		name         string
+		decisionFunc func(ids IDSet) error
+	}{
+		{
+			name: "always_succeed",
+			// Always succeed. We expect all data to be delivered as is.
+			decisionFunc: func(ids IDSet) error { return nil },
+		},
+		{
+			name:         "random_non_permanent_error",
+			decisionFunc: randomNonPermanentErrorConsumeDecision,
+		},
+		{
+			name:         "random_permanent_error",
+			decisionFunc: randomPermanentErrorConsumeDecision,
+		},
+		{
+			name:         "random_error",
+			decisionFunc: randomErrorsConsumeDecision,
+		},
+	}
+
+	for _, scenario := range scenarios {
+		params.T.Run(
+			scenario.name, func(t *testing.T) {
+				checkConsumeContractScenario(params, scenario.decisionFunc)
+			},
+		)
+	}
+}
+
+func checkConsumeContractScenario(params CheckConsumeContractParams, decisionFunc func(ids IDSet) error) {
+	consumer := &mockConsumer{t: params.T, consumeDecisionFunc: decisionFunc}
+	ctx := context.Background()
+
+	// Create and start the receiver.
+	var receiver component.Component
+	var err error
+	switch params.DataType {
+	case component.DataTypeLogs:
+		receiver, err = params.Factory.CreateLogsReceiver(ctx, NewNopCreateSettings(), params.Config, consumer)
+	case component.DataTypeTraces:
+		receiver, err = params.Factory.CreateTracesReceiver(ctx, NewNopCreateSettings(), params.Config, consumer)
+	case component.DataTypeMetrics:
+		// TODO: add metrics support to mockConsumer so that this case can be also implemented.
+		require.FailNow(params.T, "DataTypeMetrics is not implemented")
+	default:
+		require.FailNow(params.T, "must specify a valid DataType to test for")
+	}
+
+	require.NoError(params.T, err)
+
+	err = receiver.Start(ctx, componenttest.NewNopHost())
+	require.NoError(params.T, err)
+
+	// Begin generating data to the receiver.
+
+	var generatedIds IDSet
+	var generatedIndex int64
+	var mux sync.Mutex
+	var wg sync.WaitGroup
+
+	const concurrency = 4
+
+	// Create concurrent goroutines that use the generator.
+	// The total number of generator calls will be equal to params.GenerateCount.
+
+	params.Generator.Start()
+
+	for j := 0; j < concurrency; j++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for atomic.AddInt64(&generatedIndex, 1) <= int64(params.GenerateCount) {
+				ids := params.Generator.Generate()
+				require.Greater(params.T, len(ids), 0)
+
+				mux.Lock()
+				duplicates := generatedIds.mergeSlice(ids)
+				mux.Unlock()
+
+				// Check that the generator works correctly. There may not be any duplicates in the
+				// generated data set.
+				require.Empty(params.T, duplicates)
+			}
+		}()
+	}
+
+	// Wait until all generator goroutines are done.
+	wg.Wait()
+
+	// Wait until all data is seen by the consumer.
+	assert.Eventually(params.T, func() bool {
+		// Calculate the union of accepted and dropped data.
+		acceptedAndDropped, duplicates := consumer.acceptedIds.union(consumer.droppedIds)
```

Comment:
I think we can run into a read/write race condition here. `consumer.mux` should probably be used here. Or maybe better to use `sync.Map` everywhere?

Reply:
Good catch. This can indeed race. The only reason it didn't is because my test receiver doesn't do async consumption. Will fix.